types/app.go

```diff
@@ -0,0 +1,321 @@
+package sdk
+
+import (
+	"github.com/pkg/errors"
+
+	abci "github.com/tendermint/abci/types"
+	dbm "github.com/tendermint/tmlibs/db"
+	"github.com/tendermint/tmlibs/log"
+)
+
+// Key to store the header in the DB itself.
+// Use the db directly instead of a store to avoid
+// conflicts with handlers writing to the store
+// and to avoid affecting the Merkle root.
+var dbHeaderKey = []byte("header")
+
+// The ABCI application
+type App struct {
+	// initialized on creation
+	Logger     log.Logger
+	name       string           // application name from abci.Info
+	db         dbm.DB           // common DB backend
+	cms        CommitMultiStore // Main (uncached) state
+	codespacer *Codespacer      // handle module codespacing
```

Comment:
Should the codespacing logic be part of the `app` type? Codespacing is only used for modules, which this `app` type knows nothing about.

Reply:
Also, I think `app` is *most* useful as a separate abstraction for SDK clients who don't want to use separate modules or routing logic, and might not care about codespacing.